thrift/headers_test.go

```diff
@@ -41,20 +42,137 @@ var headers = map[string]string{
 	"header0": "value2",
 }
 
+var headerTests = []struct {
+	m         map[string]string
+	encoding  []byte
+	encoding2 []byte
+}{
+	{
+		m:        nil,
+		encoding: []byte{0, 0},
+	},
+	{
+		m:        make(map[string]string),
+		encoding: []byte{0, 0},
+	},
+	{
+		m: map[string]string{
+			"k": "v",
+		},
+		encoding: []byte{
+			0, 1, /* number of headers */
+			0, 1, /* length of key */
+			'k',
+			0, 1, /* length of value */
+			'v',
+		},
+	},
+	{
+		m: map[string]string{
+			"": "",
+		},
+		encoding: []byte{
+			0, 1, /* number of headers */
+			0, 0,
+			0, 0,
+		},
+	},
+	{
+		m: map[string]string{
+			"k1": "v12",
+			"k2": "v34",
+		},
+		encoding: []byte{
+			0, 2, /* number of headers */
+			0, 2, /* length of key */
+			'k', '2',
+			0, 3, /* length of value */
+			'v', '3', '4',
+			0, 2, /* length of key */
+			'k', '1',
+			0, 3, /* length of value */
+			'v', '1', '2',
+		},
+		encoding2: []byte{
+			0, 2, /* number of headers */
+			0, 2, /* length of key */
+			'k', '1',
+			0, 3, /* length of value */
+			'v', '1', '2',
+			0, 2, /* length of key */
+			'k', '2',
+			0, 3, /* length of value */
+			'v', '3', '4',
+		},
+	},
+}
+
+func TestWriteHeadersSuccessful(t *testing.T) {
+	for _, tt := range headerTests {
+		buf := &bytes.Buffer{}
+		err := WriteHeaders(buf, tt.m)
+		assert.NoError(t, err, "WriteHeaders failed")
+
+		// Writes iterate over the map in an undefined order, so we might
+		// get expected, or expected2.
+		if bytes.Equal(tt.encoding, buf.Bytes()) ||
+			bytes.Equal(tt.encoding2, buf.Bytes()) {
+			continue
+		}
+
+		assert.Equal(t, tt.encoding, buf.Bytes(), "Unexpected bytes")
+		assert.Equal(t, tt.encoding2, buf.Bytes(), "Unexpected bytes")
```

Comment:
Rewrote the logic so it's a little clearer -- do you think it's better without need for a log now?